rules/windows/execution_posh_psreflect.toml

```diff
@@ -0,0 +1,118 @@
+
+PSReflect is a library that enables PowerShell to access win32 API functions in an uncomplicated way. It also helps to
+create enums and structs easily—all without touching the disk.
+
+Although this is an interesting project for every developer and admin out there, it is mainly used in the red team and
+malware tooling for its capabilities.
+
+Detecting the core implementation of PSReflect means detecting most of the tooling that uses windows API through
+PowerShell, enabling the defender to discover tools being dropped in the environment.
+
+#### Possible investigation steps:
+- Check for additional PowerShell logs that indicate that the script/command was run.
+- Gather the script content that may be split into multiple script blocks, and identify its capabilities.
+- If the action is suspicious for the user, check for any other activities done by the user in the last 48 hours.
+- Look for additional alerts involving the host and the user.
+
+### False Positive Analysis
+- Verify whether the script content is malicious/harmful.
+
+### Related Rules
+- PowerShell Suspicious Discovery Related Windows API Functions
+- PowerShell Keylogging Script
+- PowerShell Suspicious Script with Audio Capture Capabilities
+- Potential Process Injection via PowerShell
+- PowerShell Reflection Assembly Load
+- PowerShell Suspicious Payload Encoded and Compressed
+- PowerShell Suspicious Script with Screenshot Capabilities
+
+### Response and Remediation
+- Immediate response should be taken to validate activity, investigate and potentially isolate activity to prevent further
+post-compromise behavior.
+
+## Config
+The 'PowerShell Script Block Logging' logging policy is required be configured (Enable).
+
+Steps to implement the logging policy with with Advanced Audit Configuration:
+```
+Computer Configuration > 
+Administrative Templates > 
+Windows PowerShell > 
+Turn on PowerShell Script Block Logging (Enable)
+```
+Steps to implement the logging policy via registry:
+```
+reg add "hklm\\SOFTWARE\\Policies\\Microsoft\\Windows\\PowerShell\\ScriptBlockLogging" /v EnableScriptBlockLogging /t REG_DWORD /d 1
+```
+"""
+references = [
+  "https://github.com/mattifestation/PSReflect/blob/master/PSReflect.psm1",
+  "https://github.com/atc-project/atc-data/blob/master/docs/Logging_Policies/LP_0109_windows_powershell_script_block_log.md"
+]
+risk_score = 47
+rule_id = "56f2e9b5-4803-4e44-a0a4-a52dc79d57fe"
+severity = "medium"
+tags = ["Elastic", "Host", "Windows", "Threat Detection", "Execution"]
+timestamp_override = "event.ingested"
+type = "query"
+
+query = '''
+event.code:"4104" and 
+  powershell.file.script_block_text : (
+    New-InMemoryModule or
+    Add-Win32Type or
+    psenum or
+    DefineDynamicAssembly or
+    DefineDynamicModule or
+    Reflection.TypeAttributes or
+    Reflection.Emit.OpCodes or
+    Reflection.Emit.CustomAttributeBuilder or
+    Runtime.InteropServices.DllImportAttribute
+  )
```

Comment:
This seems overly FP prone to search any _any_ one of these. Would it be safer and sufficient to detect one of the defined function names in addition to a call to one of those imports?

```
    powershell.file.script_block_text:(
      New-InMemoryModule or
      Add-Win32Type or
      psenum) and
    powershell.file.script_block_text:(
      DefineDynamicAssembly or
      DefineDynamicModule or
      Reflection.TypeAttributes or
      Reflection.Emit.OpCodes or
      Reflection.Emit.CustomAttributeBuilder or
      Runtime.InteropServices.DllImportAttribute
```

Reply:
I will elaborate the logic a little more, but I would not go with relying on the Function names...Maybe an 'and' condition on the second block

Reply:
I took a look on this, but as the Posh ScriptBlock Logs are split into different events, We would miss most of the tooling used in the wild, I think that the current query is pretty confident to spot PSReflect and similar capabilities. These are not commonly found in benign Scripts. Ref:

New-InMemoryModule: https://grep.app/search?q=New-InMemoryModule&filter[lang][0]=PowerShell
Add-Win32Type: https://grep.app/search?q=Add-Win32Type&filter[lang][0]=PowerShell
psenum: https://grep.app/search?q=psenum&filter[lang][0]=PowerShell
DefineDynamicAssembly: https://grep.app/search?q=DefineDynamicAssembly&filter[lang][0]=PowerShell
DefineDynamicModule: https://grep.app/search?q=DefineDynamicModule&filter[lang][0]=PowerShell
Reflection.TypeAttributes: https://grep.app/search?q=Reflection.TypeAttributes&filter[lang][0]=PowerShell
Reflection.Emit.OpCodes: https://grep.app/search?q=Reflection.Emit.OpCodes&filter[lang][0]=PowerShell
Reflection.Emit.CustomAttributeBuilder: https://grep.app/search?q=Reflection.Emit.CustomAttributeBuilder&filter[lang][0]=PowerShell
Runtime.InteropServices.DllImportAttribute: https://grep.app/search?q=Runtime.InteropServices.DllImportAttribute&filter[lang][0]=PowerShell

Reply:
>Posh ScriptBlock Logs are split into different events, We would miss most of the tooling used in the wild

can you elaborate on this? It still seems too loose. I also am not sure I understand how the revised logic suggestion would miss. Do you have example logs?

Reply:
> can you elaborate on this? It still seems too loose. I also am not sure I understand how the revised logic suggestion would miss. Do you have example logs?

Powershell Scriptblock Logging has a size limitation, which splits large scripts into different (event id 4104) events in a sequence, here is an example:

![image](https://user-images.githubusercontent.com/26856693/145282243-5db23b07-1242-4e2e-84ad-242e7c1f8044.png)

As we don't support text fields in EQL, there is no way of looking for the content split in different events currently, so we would miss this.

The suggested logic is a bit fragile because the attacker controls those function names, and in the query are just low-hanging fruits. Reflection and assembly methods are more adequate to catch more stuff.